src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs

```diff
@@ -0,0 +1,688 @@
+
+namespace {_generationNamespace}
+{{
+    {JsonContextDeclarationSource}
+    {{
+        private JsonTypeInfo<{typeCompilableName}> _{typeFriendlyName};
+        public JsonTypeInfo<{typeCompilableName}> {typeFriendlyName}
+        {{
+            get
+            {{
+                if (_{typeFriendlyName} == null)
+                {{
+                    {WrapWithCheckForCustomConverterIfRequired(metadataInitSource, typeCompilableName, typeFriendlyName, GetNumberHandlingAsStr(typeMetadata.NumberHandling))}
+                }}
+
+                return _{typeFriendlyName};
+            }}
+        }}{additionalSource}
+    }}
+}}
+";
+            }
+
+            private string WrapWithCheckForCustomConverterIfRequired(string source, string typeCompilableName, string typeFriendlyName, string numberHandlingNamedArg)
+            {
+                if (!_honorRuntimeProvidedCustomConverters)
+                {
+                    return source;
+                }
+
+                return @$"JsonConverter customConverter;
+                    if ({OptionsInstanceVariableName}.Converters.Count > 0 && (customConverter = {RuntimeCustomConverterFetchingMethodName}(typeof({typeCompilableName}))) != null)
+                    {{
+                        _{typeFriendlyName} = {JsonMetadataServicesClassName}.{GetCreateValueInfoMethodRef(typeCompilableName)}({OptionsInstanceVariableName}, customConverter);
+                    }}
+                    else
+                    {{
+                        {source.Replace(Environment.NewLine, $"{Environment.NewLine}    ")}
+                    }}";
+            }
+
+            // Base source generation context partial class.
+            private string BaseJsonContextImplementation()
+            {
+                StringBuilder sb = new();
+                sb.Append(@$"using System.Text.Json;
+using System.Text.Json.Serialization;
+
+namespace {_generationNamespace}
+{{
+    {JsonContextDeclarationSource}
+    {{
+        private static JsonContext s_default;
+        public static JsonContext Default => s_default ??= new JsonContext(new JsonSerializerOptions());
+
+        public JsonContext() : base(null)
+        {{
+        }}
+
+        public JsonContext(JsonSerializerOptions options) : base(options)
+        {{
+        }}
+
+        {GetFetchLogicForRuntimeSpecifiedCustomConverter()}
+    }}
+}}
+");
+
+                return sb.ToString();
+            }
+
+            private string GetFetchLogicForRuntimeSpecifiedCustomConverter()
+            {
+                if (!_honorRuntimeProvidedCustomConverters)
+                {
+                    return "";
+                }
+
+                return @$"private JsonConverter {RuntimeCustomConverterFetchingMethodName}(System.Type type)
+        {{
+            System.Collections.Generic.IList<JsonConverter> converters = {OptionsInstanceVariableName}.Converters;
+
+            // TODO: use a dictionary if count > ~15.
```

Comment:
Same question as above - do we want to generate TODOs in user source?

Also - what is the plan for tracking these TODOs?

Reply:
I'll remove the TODOs & pull them out into an issue.